Build failure: libstrophe

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

`nix-build -A libstrophe`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/318046793

### Relevant log output

```console
/nix/store/lw117lsr8d585xs63kx5k233impyrq7q-bash-5.3p3/bin/bash ./libtool  --tag=CC   --mode=compile gcc -DPACKAGE_NAME=\"libstrophe\" -DPACKAGE_TARNAME=\"libstrophe\" -DPACKAGE_VERSION=\"0.14.0\" -DPACKAGE_STRING=\"libstrophe\ 0.14.0\" -DPACKAGE_BUGREPORT=\"jack@metajack.im\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libstrophe\" -DVERSION=\"0.14.0\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_CLOCK_GETTIME=1 -DHAVE_SNPRINTF=1 -DHAVE_VSNPRINTF=1 -DHAVE_VA_COPY=1 -DHAVE_GETRANDOM=1 -DHAVE_ZLIB=1 -DLIBXMPP_VERSION_MAJOR=0 -DLIBXMPP_VERSION_MINOR=14 -I.    -I/nix/store/2ivy0r8ab3bnps5957vfrxcjfcgad661-openssl-3.6.0-dev/include -I. -Wall -Wextra -I/nix/store/59h4nhpmifmrmn8rr98aq1559rsx3r04-expat-2.7.3-dev/include   -I/nix/store/srnvjrr38j7xmpcmjpndd6mw31lsfjrl-zlib-1.3.1-dev/include -g -O2 -c -o src/libstrophe_la-handler.lo `test -f 'src/handler.c' || echo './'`src/handler.c
libtool: compile:  gcc -DPACKAGE_NAME=\"libstrophe\" -DPACKAGE_TARNAME=\"libstrophe\" -DPACKAGE_VERSION=\"0.14.0\" "-DPACKAGE_STRING=\"libstrophe 0.14.0\"" -DPACKAGE_BUGREPORT=\"jack@metajack.im\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libstrophe\" -DVERSION=\"0.14.0\" -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_CLOCK_GETTIME=1 -DHAVE_SNPRINTF=1 -DHAVE_VSNPRINTF=1 -DHAVE_VA_COPY=1 -DHAVE_GETRANDOM=1 -DHAVE_ZLIB=1 -DLIBXMPP_VERSION_MAJOR=0 -DLIBXMPP_VERSION_MINOR=14 -I. -I/nix/store/2ivy0r8ab3bnps5957vfrxcjfcgad661-openssl-3.6.0-dev/include -I. -Wall -Wextra -I/nix/store/59h4nhpmifmrmn8rr98aq1559rsx3r04-expat-2.7.3-dev/include -I/nix/store/srnvjrr38j7xmpcmjpndd6mw31lsfjrl-zlib-1.3.1-dev/include -g -O2 -c src/handler.c  -fPIC -DPIC -o src/.libs/libstrophe_la-handler.o
src/handler.c: In function 'handler_fire_stanza':
src/handler.c:92:20: warning: cast between incompatible function types from 'int (*)(void)' to 'int (*)(xmpp_conn_t *, xmpp_stanza_t *, void *)' {aka 'int (*)(struct _xmpp_conn_t *, struct _xmpp_stanza_t *, void *)'} [-Wcast-function-type]
   92 |             ret = ((xmpp_handler)(item->handler))(conn, stanza, item->userdata);
      |                    ^
src/handler.c:134:20: warning: cast between incompatible function types from 'int (*)(void)' to 'int (*)(xmpp_conn_t *, xmpp_stanza_t *, void *)' {aka 'int (*)(struct _xmpp_conn_t *, struct _xmpp_stanza_t *, void *)'} [-Wcast-function-type]
  134 |             ret = ((xmpp_handler)(item->handler))(conn, stanza, item->userdata);
      |                    ^
src/handler.c: In function 'handler_fire_timed':
src/handler.c:193:24: warning: cast between incompatible function types from 'int (*)(void)' to 'int (*)(xmpp_conn_t *, void *)' {aka 'int (*)(struct _xmpp_conn_t *, void *)'} [-Wcast-function-type]
  193 |                 ret = ((xmpp_timed_handler)item->handler)(conn, item->userdata);
      |                        ^
src/handler.c:224:18: warning: cast between incompatible function types from 'int (*)(void)' to 'int (*)(xmpp_ctx_t *, void *)' {aka 'int (*)(struct _xmpp_ctx_t *, void *)'} [-Wcast-function-type]
  224 |                 ((xmpp_global_timed_handler)item->handler)(ctx, item->userdata);
      |                  ^
src/handler.c: In function 'xmpp_timed_handler_delete':
src/handler.c:323:61: error: passing argument 3 of '_timed_handler_delete' from incompatible pointer type [-Wincompatible-pointer-types]
  323 |     _timed_handler_delete(conn->ctx, &conn->timed_handlers, handler);
      |                                                             ^~~~~~~
      |                                                             |
      |                                                             xmpp_timed_handler {aka int (*)(struct _xmpp_conn_t *, void *)}
src/handler.c:299:53: note: expected 'xmpp_void_handler' {aka 'int (*)(void)'} but argument is of type 'xmpp_timed_handler' {aka 'int (*)(struct _xmpp_conn_t *, void *)'}
  299 |                                   xmpp_void_handler handler)
      |                                   ~~~~~~~~~~~~~~~~~~^~~~~~~
src/handler.c:28:15: note: 'xmpp_void_handler' declared here
   28 | typedef int (*xmpp_void_handler)();
      |               ^~~~~~~~~~~~~~~~~
In file included from src/handler.c:24:
./strophe.h:486:15: note: 'xmpp_timed_handler' declared here
  486 | typedef int (*xmpp_timed_handler)(xmpp_conn_t *conn, void *userdata);
      |               ^~~~~~~~~~~~~~~~~~
src/handler.c: In function '_id_handler_add':
src/handler.c:337:27: warning: comparison of distinct pointer types lacks a cast [-Wcompare-distinct-pointer-types]
  337 |         if (item->handler == handler && item->userdata == userdata) {
      |                           ^~
src/handler.c:352:19: error: assignment to 'int (*)(void)' from incompatible pointer type 'xmpp_handler' {aka 'int (*)(struct _xmpp_conn_t *, struct _xmpp_stanza_t *, void *)'} [-Wincompatible-pointer-types]
  352 |     item->handler = handler;
      |                   ^
./strophe.h:505:15: note: 'xmpp_handler' declared here
  505 | typedef int (*xmpp_handler)(xmpp_conn_t *conn,
      |               ^~~~~~~~~~~~
src/handler.c: In function 'xmpp_id_handler_delete':
src/handler.c:396:27: warning: comparison of distinct pointer types lacks a cast [-Wcompare-distinct-pointer-types]
  396 |         if (item->handler == handler) {
      |                           ^~
src/handler.c: In function '_handler_add':
src/handler.c:439:27: warning: comparison of distinct pointer types lacks a cast [-Wcompare-distinct-pointer-types]
  439 |         if (item->handler == handler && item->userdata == userdata) {
      |                           ^~
src/handler.c:454:19: error: assignment to 'int (*)(void)' from incompatible pointer type 'xmpp_handler' {aka 'int (*)(struct _xmpp_conn_t *, struct _xmpp_stanza_t *, void *)'} [-Wincompatible-pointer-types]
  454 |     item->handler = handler;
      |                   ^
./strophe.h:505:15: note: 'xmpp_handler' declared here
  505 | typedef int (*xmpp_handler)(xmpp_conn_t *conn,
      |               ^~~~~~~~~~~~
src/handler.c: In function 'xmpp_handler_delete':
src/handler.c:497:27: warning: comparison of distinct pointer types lacks a cast [-Wcompare-distinct-pointer-types]
  497 |         if (item->handler == handler) {
      |                           ^~
src/handler.c: In function 'xmpp_timed_handler_add':
src/handler.c:533:58: error: passing argument 3 of '_timed_handler_add' from incompatible pointer type [-Wincompatible-pointer-types]
  533 |     _timed_handler_add(conn->ctx, &conn->timed_handlers, handler, period,
      |                                                          ^~~~~~~
      |                                                          |
      |                                                          xmpp_timed_handler {aka int (*)(struct _xmpp_conn_t *, void *)}
src/handler.c:262:50: note: expected 'xmpp_void_handler' {aka 'int (*)(void)'} but argument is of type 'xmpp_timed_handler' {aka 'int (*)(struct _xmpp_conn_t *, void *)'}
  262 |                                xmpp_void_handler handler,
      |                                ~~~~~~~~~~~~~~~~~~^~~~~~~
src/handler.c:28:15: note: 'xmpp_void_handler' declared here
   28 | typedef int (*xmpp_void_handler)();
      |               ^~~~~~~~~~~~~~~~~
./strophe.h:486:15: note: 'xmpp_timed_handler' declared here
  486 | typedef int (*xmpp_timed_handler)(xmpp_conn_t *conn, void *userdata);
      |               ^~~~~~~~~~~~~~~~~~
src/handler.c: In function 'handler_add_timed':
src/handler.c:551:58: error: passing argument 3 of '_timed_handler_add' from incompatible pointer type [-Wincompatible-pointer-types]
  551 |     _timed_handler_add(conn->ctx, &conn->timed_handlers, handler, period,
      |                                                          ^~~~~~~
      |                                                          |
      |                                                          xmpp_timed_handler {aka int (*)(struct _xmpp_conn_t *, void *)}
src/handler.c:262:50: note: expected 'xmpp_void_handler' {aka 'int (*)(void)'} but argument is of type 'xmpp_timed_handler' {aka 'int (*)(struct _xmpp_conn_t *, void *)'}
  262 |                                xmpp_void_handler handler,
      |                                ~~~~~~~~~~~~~~~~~~^~~~~~~
src/handler.c:28:15: note: 'xmpp_void_handler' declared here
   28 | typedef int (*xmpp_void_handler)();
      |               ^~~~~~~~~~~~~~~~~
./strophe.h:486:15: note: 'xmpp_timed_handler' declared here
  486 | typedef int (*xmpp_timed_handler)(xmpp_conn_t *conn, void *userdata);
      |               ^~~~~~~~~~~~~~~~~~
src/handler.c: In function 'xmpp_global_timed_handler_add':
src/handler.c:750:51: error: passing argument 3 of '_timed_handler_add' from incompatible pointer type [-Wincompatible-pointer-types]
  750 |     _timed_handler_add(ctx, &ctx->timed_handlers, handler, period, userdata, 1);
      |                                                   ^~~~~~~
      |                                                   |
      |                                                   xmpp_global_timed_handler {aka int (*)(struct _xmpp_ctx_t *, void *)}
src/handler.c:262:50: note: expected 'xmpp_void_handler' {aka 'int (*)(void)'} but argument is of type 'xmpp_global_timed_handler' {aka 'int (*)(struct _xmpp_ctx_t *, void *)'}
  262 |                                xmpp_void_handler handler,
      |                                ~~~~~~~~~~~~~~~~~~^~~~~~~
src/handler.c:28:15: note: 'xmpp_void_handler' declared here
   28 | typedef int (*xmpp_void_handler)();
      |               ^~~~~~~~~~~~~~~~~
./strophe.h:495:15: note: 'xmpp_global_timed_handler' declared here
  495 | typedef int (*xmpp_global_timed_handler)(xmpp_ctx_t *ctx, void *userdata);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~
src/handler.c: In function 'xmpp_global_timed_handler_delete':
src/handler.c:763:54: error: passing argument 3 of '_timed_handler_delete' from incompatible pointer type [-Wincompatible-pointer-types]
  763 |     _timed_handler_delete(ctx, &ctx->timed_handlers, handler);
      |                                                      ^~~~~~~
      |                                                      |
      |                                                      xmpp_global_timed_handler {aka int (*)(struct _xmpp_ctx_t *, void *)}
src/handler.c:299:53: note: expected 'xmpp_void_handler' {aka 'int (*)(void)'} but argument is of type 'xmpp_global_timed_handler' {aka 'int (*)(struct _xmpp_ctx_t *, void *)'}
  299 |                                   xmpp_void_handler handler)
      |                                   ~~~~~~~~~~~~~~~~~~^~~~~~~
src/handler.c:28:15: note: 'xmpp_void_handler' declared here
   28 | typedef int (*xmpp_void_handler)();
      |               ^~~~~~~~~~~~~~~~~
./strophe.h:495:15: note: 'xmpp_global_timed_handler' declared here
  495 | typedef int (*xmpp_global_timed_handler)(xmpp_ctx_t *ctx, void *userdata);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~
make: *** [Makefile:1744: src/libstrophe_la-handler.lo] Error 1
```

### Additional context

Possibly related https://github.com/NixOS/nixpkgs/issues/475479

### System metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.18.2, NixOS, 26.05 (Yarara), 26.05.20251228.c0b0e0f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2+1`
 - nixpkgs: `/nix/store/72ym7zz89wa91k3g4vz2g97pma1n51zy-xzl5dhddgjxp2p18h5qpkhvlkp8128j1-source`
```

### Notify maintainers

@devhell @flosse 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

https://gist.githubusercontent.com/toastal/dd85b83f03f0cd5a468d43a9169061dd/raw/a664538ec1c9fc288562d792bc4b44f153f75013/ms-gh-foot.md

--- a/pkgs/by-name/li/libstrophe/package.nix
+++ b/pkgs/by-name/li/libstrophe/package.nix
@@ -22,10 +22,17 @@ stdenv.mkDerivation rec {
     hash = "sha256-53O8hHyw9y0Bzs+BpGouAxuSGJxh6NSNNWZqi7RHAsY=";
   };
 
+  patches = [
+    # Newer GCC rejects implicitly weak-typed pointer casting.
+    # Upstream PR: https://github.com/strophe/libstrophe/pull/267
+    ./pointer-cast.patch
+  ];
+
   nativeBuildInputs = [
     autoreconfHook
     pkg-config
   ];
+
   buildInputs = [
     openssl
     expat

--- a/pkgs/by-name/li/libstrophe/pointer-cast.patch
+++ b/pkgs/by-name/li/libstrophe/pointer-cast.patch
@@ -1,0 +1,65 @@
+diff --git a/src/handler.c b/src/handler.c
+index 1c9bf9f..168df75 100644
+--- a/src/handler.c
++++ b/src/handler.c
+@@ -320,7 +320,7 @@ static void _timed_handler_delete(xmpp_ctx_t *ctx,
+  */
+ void xmpp_timed_handler_delete(xmpp_conn_t *conn, xmpp_timed_handler handler)
+ {
+-    _timed_handler_delete(conn->ctx, &conn->timed_handlers, handler);
++    _timed_handler_delete(conn->ctx, &conn->timed_handlers, (xmpp_void_handler)handler);
+ }
+ 
+ static void _id_handler_add(xmpp_conn_t *conn,
+@@ -349,7 +349,7 @@ static void _id_handler_add(xmpp_conn_t *conn,
+         return;
+ 
+     item->user_handler = user_handler;
+-    item->handler = handler;
++    item->handler = (xmpp_void_handler)handler;
+     item->userdata = userdata;
+     item->enabled = 0;
+     item->next = NULL;
+@@ -451,7 +451,7 @@ static void _handler_add(xmpp_conn_t *conn,
+ 
+     memset(item, 0, sizeof(*item));
+     item->user_handler = user_handler;
+-    item->handler = handler;
++    item->handler = (xmpp_void_handler)handler;
+     item->userdata = userdata;
+ 
+     if (_dup_string(conn->ctx, ns, &item->u.ns))
+@@ -530,7 +530,7 @@ void xmpp_timed_handler_add(xmpp_conn_t *conn,
+                             unsigned long period,
+                             void *userdata)
+ {
+-    _timed_handler_add(conn->ctx, &conn->timed_handlers, handler, period,
++    _timed_handler_add(conn->ctx, &conn->timed_handlers, (xmpp_void_handler)handler, period,
+                        userdata, 1);
+ }
+ 
+@@ -548,7 +548,7 @@ void handler_add_timed(xmpp_conn_t *conn,
+                        unsigned long period,
+                        void *userdata)
+ {
+-    _timed_handler_add(conn->ctx, &conn->timed_handlers, handler, period,
++    _timed_handler_add(conn->ctx, &conn->timed_handlers, (xmpp_void_handler)handler, period,
+                        userdata, 0);
+ }
+ 
+@@ -747,7 +747,7 @@ void xmpp_global_timed_handler_add(xmpp_ctx_t *ctx,
+                                    unsigned long period,
+                                    void *userdata)
+ {
+-    _timed_handler_add(ctx, &ctx->timed_handlers, handler, period, userdata, 1);
++    _timed_handler_add(ctx, &ctx->timed_handlers, (xmpp_void_handler)handler, period, userdata, 1);
+ }
+ 
+ /** Delete a global timed handler.
+@@ -760,5 +760,5 @@ void xmpp_global_timed_handler_add(xmpp_ctx_t *ctx,
+ void xmpp_global_timed_handler_delete(xmpp_ctx_t *ctx,
+                                       xmpp_global_timed_handler handler)
+ {
+-    _timed_handler_delete(ctx, &ctx->timed_handlers, handler);
++    _timed_handler_delete(ctx, &ctx->timed_handlers, (xmpp_void_handler)handler);
+ }

